File: Perl/3.pl
#!/usr/bin/perl
# AdventOfCode2020
# --- Day 3: Toboggan Trajectory --- 
#part1: 214
#part1 took 0.00085679602622 ms to execute
#part2: 8336352024
#94 214 99 91 46
#part2 took 0.01015240693092 ms to execute

use strict;
use warnings;

use Time::HiRes qw( time );
use Data::Dumper;

use constant FALSE => 0;
use constant TRUE  => 1;
use constant TOTAL => 150;

my $start  = time();
my %points = ();
my $row = 0;
my $col = 0;


my $p1 = part1(3,1);
#print "\$col:$col|\$row:$row\n";
print "part1: $p1\n";
my $end     = time();
my $runtime = sprintf( "%.16s", ( $end - $start ) / 1000 );
print "part1 took $runtime ms to execute\n\n";

part2();
$end     = time();
$runtime = sprintf( "%.16s", ( $end - $start ) / 1000 );
print "part2 took $runtime ms to execute\n";

exit(0);

sub part1 {
  my ($a,$b) = @_;

  my $counter = 0;
  %points = ();
  #my $p = load_data(\%points);
  load_data();
  #%points = %$p;
  #print "$row:$row\n";

  my $x=0;
  my $y=0;
  my $key="";
  while ($y<=$row) {
    $x=$x+$a;
    $y=$y+$b;
    $key = "($x" . "," . "$y)";
    if (defined $points{$key}) {
      if ($points{$key} eq "#") {
        $points{$key}="X";
        $counter++
      } else {
        $points{$key}="O";
      }
    }
    
  }
  return $counter;
}

sub part2 {
  # Code goes here
  my $p1 = part1(1,1);
  my $p2 = part1(3,1);
  my $p3 = part1(5,1);
  my $p4 = part1(7,1);
  my $p5 = part1(1,2);
  my $part2 = $p1 * $p2 * $p3 * $p4 * $p5;
  print "part2: $part2\n";
  print "$p1 $p2 $p3 $p4 $p5\n";
}


sub print_state {
  #my ($p) = @_;
  #my %points = %$p;
  for my $i (0..$row-1) {
    for my $j (0..$col-1) {
      my $key = "($j" . "," . "$i)";
      print $points{$key};
    }
    print "\n";
  }
  print "\n";
}


sub load_data {
  #my ($p) = @_;
  #my %points = %$p;
  $col=0;
  $row=0;
  ##### Load Data #####
  my $filename = '../data/google-3.txt';
  #my $filename = '../data/reddit-3.txt';
  # Repeat to the right 4 times.
  open( my $fh, '<:encoding(UTF-8)', $filename ) or die "Could not open file '$filename' $!";
  #while (<DATA>) {
  while (<$fh>) {
    chomp;

    $_="$_"x100;

    my @data = split '', $_;
    for my $data (@data) {
        my $colrow = "($col" . "," . "$row)";
        if ($data eq '#') {
            $points{$colrow}='#';
        } else {
            $points{$colrow}='.';
        }
        $col += 1;
    }
    $col = 0;
    $row += 1;
  }
  close $fh;
  #return \%points;
}


__DATA__
..##.........##.........##.........##.........##.......
#...#...#..#...#...#..#...#...#..#...#...#..#...#...#..
.#....#..#..#....#..#..#....#..#..#....#..#..#....#..#.
..#.#...#.#..#.#...#.#..#.#...#.#..#.#...#.#..#.#...#.#
.#...##..#..#...##..#..#...##..#..#...##..#..#...##..#.
..#.##.......#.##.......#.##.......#.##.......#.##.....
.#.#.#....#.#.#.#....#.#.#.#....#.#.#.#....#.#.#.#....#
.#........#.#........#.#........#.#........#.#........#
#.##...#...#.##...#...#.##...#...#.##...#...#.##...#...
#...##....##...##....##...##....##...##....##...##....#
.#..#...#.#.#..#...#.#.#..#...#.#.#..#...#.#.#..#...#.#

File: JavaScript/3.js
// AdventOfCode2020
// --- Day 2: Password Philosophy ---
// part1: 600
// part1: 6.45ms
// part2: 456
// part2: 1.811ms

let fs = require("fs");

function part1() {
    let data = new Array();
    fs.readFileSync("../data/google-3.txt", "utf-8")
    //fs.readFileSync("./data.txt", "utf-8")
        .split("\n")
        .forEach((line,index,arr) => {
            if (index === arr.length - 1 && line === "") { return; }
            data.push(line)
        });

    // Solve the problem here
    let ans = 1;
    let r   = 0;
    let c   = 0;
    while (r+1 < data.length) {
        c += 3
        r += 1
        if (data[r][c%data[r].length]
    }
}


import fileinput

slopes = [(1,1),(3,1),(5,1),(7,1),(1,2)]

G=[]
for line in fileinput.input():
    G.append(list(line.strip()))

ans = 1
for (dc,dr) in slopes:
    r = 0
    c = 0
    score = 0
    while r+1 < len(G):
        c+= dc
        r+= dr
        if G[r][c%len(G[r])]=="#":
            score += 1
    ans *= score
    if dc==3 and dr==1:
        print(score)
print(ans)





function part2() {
    let data = new Array();
    fs.readFileSync("../data/google-2.txt", "utf-8")
    //fs.readFileSync("./data.txt", "utf-8")
        .split("\n")
        .forEach((line,index,arr) => {
            if (index === arr.length - 1 && line === "") { return; }
            //console.log("line="+line);
            let rx_extract = /(\d{1,2})-(\d{1,2}) (\w): (\w+)/;
            m=rx_extract.exec(line)
            //console.log(m[1],m[2],m[3],m[4]);
            let row = new Array(m[1],m[2],m[3],m[4])
            data.push(row);
        });

    // Solve the problem here
    let valid = 0;
    for (i = 0; i < data.length; i++) {
        let array = data[i];
        //console.log("cnt= "+cnt,"array[1]= "+Number(array[0]),"array[2]= "+Number(array[1]));
        if (array[3].charAt(Number(array[0]))===array[2] ^ array[3].charAt(Number(array[1]))===array[2])
            valid++;
    }
    console.log("part2: " + valid);
}

console.time('part1');
part1();
console.timeEnd('part1');

console.time('part2');
part2();
console.timeEnd('part2');
